server/entitymanager/entitymanager.go

```diff
@@ -2,30 +2,86 @@
 
 import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+
 	"github.com/openconfig/bootz/proto/bootz"
 	"github.com/openconfig/bootz/server/service"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
 )
 
-type InMemoryEntityManager struct{}
+type InMemoryEntityManager struct {
+	// A mapping of chassis or control card serial number to ChassisEntity.
+	inventory map[string]*service.ChassisEntity
+}
 
-func (m *InMemoryEntityManager) ResolveChassis(lookup *service.EntityLookup) (*service.ChassisEntity, error) {
-	return nil, status.Errorf(codes.Unimplemented, "Unimplemented")
+func (m *InMemoryEntityManager) ResolveChassis(desc *bootz.ChassisDescriptor) (*service.ChassisEntity, error) {
+	// Attempt to resolve a fixed form factor chassis.
+	if desc.GetSerialNumber() != "" {
+		c, ok := m.inventory[desc.GetSerialNumber()]
+		if !ok {
+			return nil, status.Errorf(codes.NotFound, "fixed form factor chassis %v not found in inventory", desc.GetSerialNumber())
+		}
+		return c, nil
+	}
+	// Attempt to resolve control cards to chassis.
+	for _, card := range desc.GetControlCards() {
+		c, ok := m.inventory[card.GetSerialNumber()]
+		if !ok {
+			continue
+		}
+		return c, nil
+	}
+	return nil, status.Errorf(codes.NotFound, "chassis descriptor %v not found in inventory", desc)
 }
 
 func (m *InMemoryEntityManager) GetBootstrapData(*bootz.ControlCard) (*bootz.BootstrapDataResponse, error) {
 	return nil, status.Errorf(codes.Unimplemented, "Unimplemented")
 }
 
-func (m *InMemoryEntityManager) SetStatus(*bootz.ReportStatusRequest) error {
+func (m *InMemoryEntityManager) SetStatus(req *bootz.ReportStatusRequest) error {
 	return status.Errorf(codes.Unimplemented, "Unimplemented")
 }
 
-func (m *InMemoryEntityManager) Sign(*bootz.GetBootstrapDataResponse) error {
-	return status.Errorf(codes.Unimplemented, "Unimplemented")
+// Sign unmarshals the SignedResponse bytes then generates a signature from its Ownership Certificate private key.
+func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, priv *rsa.PrivateKey) error {
+	if resp.GetSignedResponse() == nil {
+		return status.Errorf(codes.InvalidArgument, "empty signed response")
+	}
+	signedResponseBytes, err := proto.Marshal(resp.GetSignedResponse())
+	if err != nil {
+		return err
+	}
+	hashed := sha256.Sum256(signedResponseBytes)
+	sig, err := rsa.SignPKCS1v15(nil, priv, crypto.SHA256, hashed[:])
+	if err != nil {
+		return err
+	}
+	resp.ResponseSignature = string(sig)
+	return nil
 }
 
 func New() *InMemoryEntityManager {
-	return &InMemoryEntityManager{}
+	// TODO: Populate these values from a config file or command line flag.
+	// This represents a fixed form factor chassis (e.g. no control cards).
+	fffChassis := service.ChassisEntity{
+		BootMode: "SecureOnly",
+	}
+	// This represents a modular chassis (e.g. contains control cards).
+	modularChassis := service.ChassisEntity{
+		BootMode: "SecureOnly",
+	}
```

Comment:
is this separation required if they are same?

Reply:
I guess not. Done.